torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -260,8 +344,9 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
     // struct are empty, and there is no pre-existing accumulation tensor.
     // Directly assign the sparse tensor to the `contents` field.
     replica.contents = grad;
-    // The grad is not modified and dosesn't need to be written back.
-    return false;
+    replica.contents.div_(process_group_->getSize());
+    // The grad is modified in place and needs to be written back.
+    return true;
```

Comment:
@mrshenli are the diffs here ok? In `mark_variable_ready_dense` mul_out averages the pre-allreduce gradients (which doesn't affect grad itself, because the result is written to the allreduce bucket).  For `mark_variable_ready_sparse` to mirror that control flow, I added `div_` here.  For sparse gradients, replica.contents = grad, though, so grad itself IS affected in place.  Therefore, I also changed `return false` to `return true`.  Admittedly I don't fully understand the implications of doing so.

Reply:
Maybe the `div_`+`return true` here actually fixes a bug with the `dist_autograd` changes.  In current master, the bucket predivision is carried out in `mark_variable_ready`.  For sparse gradients, the predivision alters the grad in place, but there's no mention at that point of `runGradCallbackForVariable`, so that change isn't communicated anywhere.  Again, admittedly i have no idea what's going on.

Reply:
IIUC, the criteria for returning true is when we want to trigger cbs on that grad. So the prior (return true when the grad is modified) comments might not be accurate. 

I want to trigger those cbs when the grad is ready, i.e., after the allreduce sync done by DDP. So I think we should still return false here?

@pritamdamania87 please correct me if I am wrong.

Reply:
I think for consistency with local autograd its better to return true here (since for local autograd .grad would change, so for dist_autograd grads in the map should change too). Although, is it possible to do the division in the end in `finalize_backward`? That way we modify the grad in only one place?

Reply:
Wait a minute, since for sparse grads the grad is the bucket, it makes no sense for anything external (local or remote) to look at its values between when the allreduce hook triggers and when until the allreduce is finalized.  From the perspective of external code in this thread, the allreduce hook carries out a `div_` and immediately kicks off an allreduce which spends a while updating grad's values in place.  Any access to the grad before `finalize_backward` is racy.

We could make `div_` an out-of-place `div` for sparse gradients.  Then `variable.grad` itself will remain untouched until `finalize_bucket_sparse` (which makes sparse grad control flow more consistent with dense grads which use separate memory for their allreduce buckets).  However, as I said earlier, `finalize_bucket_sparse` makes no reference to `runGradCallbackForVariable` anyway, so `dist_autograd`'s "map" is never informed even after the allreduce completes and grad is safe to access.  Do we need to add `runGradCallbackForVariable` to `finalize_bucket_sparse`?

> is it possible to do the division in the end in finalize_backward

That's a reasonable idea but I'm wary of it.  For some networks that allreduce FP16 gradients at scale, we've noticed that post-allreduce averaging caused nonconvergence, but pre-allreduce averaging was fine.  I'm not aware off the top of my head of any cases where we observed the reverse, so I prefer the predivision that DDP implements currently.

Reply:
@mcarilli My bad I misread this piece of code initially. We should actually return false here instead of true since we are not actually modifying `grad` here. `runGradCallbackForVariable` should return True only if we change the gradients for the variable.

> Do we need to add runGradCallbackForVariable to finalize_bucket_sparse?

Yes we do and this was probably a bug in #37998. This probably wasn't caught there since we didn't have a test for sparse gradients with distributed autograd. Although, I feel we probably should fix that as a separate PR with good test coverage instead of adding it here :)

Reply:
Yes, it is true that nothing should access the grad between allreduce hooks and finalize_backward. Although, my point was mostly from a consistency standpoint. In the local autograd engine case, we are modifying the grad here and technically anyone looking at variable's .grad after the `div_` operation would see a change. So if we're updating .grad here, we should update it in distributed autograd's map as well. The other option is not to update in both .grad and dist autograd map. Although, I'd prefer the in place `div_` to avoid creating a new tensor.

> Do we need to add runGradCallbackForVariable to finalize_bucket_sparse?

Yes, we do and this was a bug in  #37998.